File: agents/github-action-agent/src/tonic_agent/github_api.py
"""Minimal GitHub REST helpers (stdlib urllib)."""

from __future__ import annotations

import base64
import json
import os
import urllib.error
import urllib.request
from typing import Any


def _api_root() -> str:
    return os.environ.get("GITHUB_API_URL", "https://api.github.com").rstrip("/")


def fetch_url_text_authenticated(url: str, token: str, *, timeout: int = 120) -> str | None:
    """GET arbitrary URL with GitHub token (used for Contents API download_url)."""
    req = urllib.request.Request(
        url,
        headers={
            "Accept": "application/vnd.github.raw",
            "Authorization": f"Bearer {token}",
            "X-GitHub-Api-Version": "2022-11-28",
        },
        method="GET",
    )
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            raw = resp.read()
            if b"\x00" in raw:
                return None
            try:
                return raw.decode("utf-8")
            except UnicodeDecodeError:
                return raw.decode("utf-8", errors="replace")
    except urllib.error.HTTPError:
        return None


def get_git_blob_text(owner: str, repo: str, blob_sha: str, token: str) -> str | None:
    """Decode UTF-8 text from GET /git/blobs/{sha} (handles large files without inline Contents)."""
    api = _api_root()
    url = f"{api}/repos/{owner}/{repo}/git/blobs/{blob_sha}"
    try:
        data = get_json(url, token)
    except RuntimeError:
        return None
    if not isinstance(data, dict):
        return None
    enc = data.get("encoding")
    content = data.get("content")
    if enc != "base64" or not isinstance(content, str):
        return None
    raw = base64.b64decode(content)
    if b"\x00" in raw:
        return None
    try:
        return raw.decode("utf-8")
    except UnicodeDecodeError:
        return raw.decode("utf-8", errors="replace")


def get_json(url: str, token: str) -> dict[str, Any] | list[Any]:
    req = urllib.request.Request(
        url,
        headers={
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {token}",
            "X-GitHub-Api-Version": "2022-11-28",
        },
        method="GET",
    )
    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            raw = resp.read().decode("utf-8")
            return json.loads(raw) if raw else {}
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"GitHub API {e.code}: {e.read().decode('utf-8', errors='replace')}") from e


def _request(
    method: str,
    url: str,
    token: str,
    body: dict[str, Any] | None = None,
) -> dict[str, Any] | list[Any]:
    data = None if body is None else json.dumps(body).encode("utf-8")
    req = urllib.request.Request(
        url,
        data=data,
        headers={
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {token}",
            "X-GitHub-Api-Version": "2022-11-28",
        },
        method=method,
    )
    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            raw = resp.read().decode("utf-8")
            return json.loads(raw) if raw else {}
    except urllib.error.HTTPError as e:
        raise RuntimeError(f"GitHub API {e.code}: {e.read().decode('utf-8', errors='replace')}") from e


def list_issue_comments(owner: str, repo: str, issue_number: int, token: str) -> list[dict[str, Any]]:
    api = _api_root()
    out: list[dict[str, Any]] = []
    page = 1
    while True:
        url = f"{api}/repos/{owner}/{repo}/issues/{issue_number}/comments?page={page}&per_page=100"
        data = get_json(url, token)
        if not isinstance(data, list) or not data:
            break
        out.extend(data)
        if len(data) < 100:
            break
        page += 1
    return out


def update_issue_comment(owner: str, repo: str, comment_id: int, body: str, token: str) -> dict[str, Any]:
    api = _api_root()
    url = f"{api}/repos/{owner}/{repo}/issues/comments/{comment_id}"
    return _request("PATCH", url, token, {"body": body})  # type: ignore[return-value]


def post_issue_comment(owner: str, repo: str, issue_number: int, body: str, token: str) -> dict[str, Any]:
    api = _api_root()
    url = f"{api}/repos/{owner}/{repo}/issues/{issue_number}/comments"
    return _request("POST", url, token, {"body": body})  # type: ignore[return-value]


def post_pull_review_comment(
    owner: str,
    repo: str,
    pull_number: int,
    body: str,
    commit_id: str,
    path: str,
    line: int,
    side: str,
    token: str,
    *,
    start_line: int | None = None,
    start_side: str | None = None,
    in_reply_to: int | None = None,
) -> dict[str, Any]:
    api = _api_root()
    url = f"{api}/repos/{owner}/{repo}/pulls/{pull_number}/comments"
    payload: dict[str, Any] = {
        "body": body,
        "commit_id": commit_id,
        "path": path,
        "line": line,
        "side": side,
    }
    if start_line is not None and start_line < line:
        payload["start_line"] = start_line
        payload["start_side"] = start_side or side
    if in_reply_to is not None:
        payload["in_reply_to"] = in_reply_to
    return _request("POST", url, token, payload)  # type: ignore[return-value]


def list_pull_review_comments(owner: str, repo: str, pull_number: int, token: str) -> list[dict[str, Any]]:
    api = _api_root()
    out: list[dict[str, Any]] = []
    page = 1
    while True:
        url = f"{api}/repos/{owner}/{repo}/pulls/{pull_number}/comments?page={page}&per_page=100"
        data = get_json(url, token)
        if not isinstance(data, list) or not data:
            break
        out.extend(data)
        if len(data) < 100:
            break
        page += 1
    return out


def upsert_issue_comment(
    owner: str,
    repo: str,
    issue_number: int,
    token: str,
    marker_prefix: str,
    body: str,
) -> dict[str, Any]:
    for c in list_issue_comments(owner, repo, issue_number, token):
        existing = c.get("body") or ""
        if marker_prefix in existing:
            cid = c.get("id")
            if isinstance(cid, int):
                return update_issue_comment(owner, repo, cid, body, token)
    return post_issue_comment(owner, repo, issue_number, body, token)


def create_pull_request(
    owner: str,
    repo: str,
    token: str,
    *,
    title: str,
    head: str,
    base: str,
    body: str,
) -> tuple[int, str]:
    api = _api_root()
    url = f"{api}/repos/{owner}/{repo}/pulls"
    data = _request(
        "POST",
        url,
        token,
        {"title": title, "head": head, "base": base, "body": body},
    )
    if not isinstance(data, dict):
        raise RuntimeError("create_pull_request: invalid response payload")
    number = data.get("number")
    head_obj = data.get("head")
    if not isinstance(number, int) or not isinstance(head_obj, dict):
        raise RuntimeError("create_pull_request: missing number/head")
    head_sha = head_obj.get("sha")
    if not isinstance(head_sha, str) or not head_sha:
        raise RuntimeError("create_pull_request: missing head sha")
    return number, head_sha

File: agents/github-action-agent-node/src/githubApi.ts
export const apiRoot = (): string =>
  (process.env.GITHUB_API_URL ?? "https://api.github.com").replace(/\/$/, "");

export const ghHeaders = (token: string): Record<string, string> => ({
  Accept: "application/vnd.github+json",
  Authorization: `Bearer ${token}`,
  "X-GitHub-Api-Version": "2022-11-28",
});

export async function githubGetJson(url: string, token: string): Promise<unknown> {
  const res = await fetch(url, { headers: ghHeaders(token) });
  const text = await res.text();
  if (!res.ok) {
    throw new Error(`GitHub API ${res.status}: ${text}`);
  }
  return text ? JSON.parse(text) : {};
}

export async function githubRequestJson(
  method: string,
  url: string,
  token: string,
  body?: Record<string, unknown>,
): Promise<unknown> {
  const res = await fetch(url, {
    method,
    headers: {
      ...ghHeaders(token),
      "Content-Type": "application/json",
    },
    body: body ? JSON.stringify(body) : undefined,
  });
  const text = await res.text();
  if (!res.ok) {
    throw new Error(`GitHub API ${res.status}: ${text}`);
  }
  return text ? JSON.parse(text) : {};
}

export async function fetchUrlTextAuthenticated(url: string, token: string): Promise<string | null> {
  try {
    const res = await fetch(url, {
      headers: {
        Accept: "application/vnd.github.raw",
        Authorization: `Bearer ${token}`,
        "X-GitHub-Api-Version": "2022-11-28",
      },
    });
    if (!res.ok) {
      return null;
    }
    const buf = Buffer.from(await res.arrayBuffer());
    if (buf.includes(0)) {
      return null;
    }
    return buf.toString("utf8");
  } catch {
    return null;
  }
}

export async function getGitBlobText(owner: string, repo: string, blobSha: string, token: string): Promise<string | null> {
  try {
    const url = `${apiRoot()}/repos/${owner}/${repo}/git/blobs/${blobSha}`;
    const data = (await githubGetJson(url, token)) as Record<string, unknown>;
    if (data.encoding !== "base64" || typeof data.content !== "string") {
      return null;
    }
    const raw = Buffer.from(data.content, "base64");
    if (raw.includes(0)) {
      return null;
    }
    return raw.toString("utf8");
  } catch {
    return null;
  }
}

export async function postIssueComment(
  owner: string,
  repo: string,
  issueNumber: number,
  body: string,
  token: string,
): Promise<unknown> {
  const url = `${apiRoot()}/repos/${owner}/${repo}/issues/${issueNumber}/comments`;
  return githubRequestJson("POST", url, token, { body });
}

export async function listIssueComments(
  owner: string,
  repo: string,
  issueNumber: number,
  token: string,
): Promise<Array<{ id?: number; body?: string }>> {
  const out: Array<{ id?: number; body?: string }> = [];
  let page = 1;
  for (;;) {
    const url = `${apiRoot()}/repos/${owner}/${repo}/issues/${issueNumber}/comments?page=${page}&per_page=100`;
    const data = (await githubGetJson(url, token)) as unknown;
    if (!Array.isArray(data) || data.length === 0) {
      break;
    }
    out.push(...(data as Array<{ id?: number; body?: string }>));
    if (data.length < 100) {
      break;
    }
    page += 1;
  }
  return out;
}

export async function updateIssueComment(
  owner: string,
  repo: string,
  commentId: number,
  body: string,
  token: string,
): Promise<unknown> {
  const url = `${apiRoot()}/repos/${owner}/${repo}/issues/comments/${commentId}`;
  return githubRequestJson("PATCH", url, token, { body });
}

export async function upsertIssueComment(
  owner: string,
  repo: string,
  issueNumber: number,
  token: string,
  markerPrefix: string,
  body: string,
): Promise<unknown> {
  const comments = await listIssueComments(owner, repo, issueNumber, token);
  const hit = comments.find((c) => (c.body ?? "").includes(markerPrefix));
  if (hit?.id != null) {
    return updateIssueComment(owner, repo, hit.id, body, token);
  }
  return postIssueComment(owner, repo, issueNumber, body, token);
}

export async function listPullReviewComments(
  owner: string,
  repo: string,
  pullNumber: number,
  token: string,
): Promise<Array<{ body?: string }>> {
  const out: Array<{ body?: string }> = [];
  let page = 1;
  for (;;) {
    const url = `${apiRoot()}/repos/${owner}/${repo}/pulls/${pullNumber}/comments?page=${page}&per_page=100`;
    const data = (await githubGetJson(url, token)) as unknown;
    if (!Array.isArray(data) || data.length === 0) {
      break;
    }
    out.push(...(data as Array<{ body?: string }>));
    if (data.length < 100) {
      break;
    }
    page += 1;
  }
  return out;
}

export async function postPullReviewComment(
  owner: string,
  repo: string,
  pullNumber: number,
  body: string,
  commitId: string,
  path: string,
  line: number,
  side: "LEFT" | "RIGHT",
  token: string,
  opts?: { startLine?: number; startSide?: string; inReplyTo?: number },
): Promise<unknown> {
  const url = `${apiRoot()}/repos/${owner}/${repo}/pulls/${pullNumber}/comments`;
  const payload: Record<string, unknown> = {
    body,
    commit_id: commitId,
    path,
    line,
    side,
  };
  if (opts?.startLine != null && opts.startLine < line) {
    payload.start_line = opts.startLine;
    payload.start_side = opts.startSide ?? side;
  }
  if (opts?.inReplyTo != null) {
    payload.in_reply_to = opts.inReplyTo;
  }
  return githubRequestJson("POST", url, token, payload);
}

export async function createPullRequest(
  owner: string,
  repo: string,
  token: string,
  params: {
    title: string;
    head: string;
    base: string;
    body: string;
  },
): Promise<{ number: number; headSha: string }> {
  const url = `${apiRoot()}/repos/${owner}/${repo}/pulls`;
  const data = (await githubRequestJson("POST", url, token, {
    title: params.title,
    head: params.head,
    base: params.base,
    body: params.body,
  })) as Record<string, unknown>;
  const number = data.number;
  const head = data.head as Record<string, unknown> | undefined;
  const headSha = head?.sha;
  if (typeof number !== "number" || typeof headSha !== "string") {
    throw new Error("createPullRequest returned incomplete payload");
  }
  return { number, headSha };
}
